File: backend/registration-service/src/main/kotlin/com/playsay/registration/service/PasswordPolicy.kt
package com.playsay.registration.service

import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.server.ResponseStatusException

@Component
class PasswordPolicy {
    fun requireValid(password: String, email: String, displayName: String? = null) {
        val failures = validate(password = password, email = email, displayName = displayName)
        if (failures.isNotEmpty()) {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, failures.joinToString(","))
        }
    }

    fun validate(password: String, email: String, displayName: String? = null): List<String> {
        val failures = mutableListOf<String>()
        val lowered = password.lowercase()

        if (password.length < minLength) {
            failures += "PASSWORD_TOO_SHORT"
        }
        if (password.length > maxLength) {
            failures += "PASSWORD_TOO_LONG"
        }
        if (weakFragments.any { lowered.contains(it) }) {
            failures += "PASSWORD_TOO_COMMON"
        }
        if (email.substringBefore("@").normalizedFragment()?.let { lowered.contains(it) } == true) {
            failures += "PASSWORD_CONTAINS_EMAIL"
        }
        if (displayName?.normalizedNameFragments()?.any { lowered.contains(it) } == true) {
            failures += "PASSWORD_CONTAINS_NAME"
        }
        if (password.characterClassCount() < minimumCharacterClasses) {
            failures += "PASSWORD_NEEDS_VARIETY"
        }

        return failures
    }

    private fun String.characterClassCount(): Int =
        listOf(
            any { it.isLowerCase() },
            any { it.isUpperCase() },
            any { it.isDigit() },
            any { !it.isLetterOrDigit() },
        ).count { it }

    private fun String.normalizedFragment(): String? =
        filter { it.isLetterOrDigit() }
            .lowercase()
            .takeIf { it.length >= 3 }

    private fun String.normalizedNameFragments(): List<String> =
        split(Regex("[^\\p{L}\\p{N}]+"))
            .mapNotNull { it.normalizedFragment() }

    private companion object {
        const val minLength = 8
        const val maxLength = 128
        const val minimumCharacterClasses = 3
        val weakFragments = setOf("password", "qwerty", "12345678", "letmein", "admin", "playsay", "play-and-say")
    }
}

File: frontend/web-app/src/features/registration/model/passwordPolicy.ts
export type PasswordIssue =
  | "tooShort"
  | "tooLong"
  | "tooCommon"
  | "containsEmail"
  | "containsName"
  | "needsVariety";

export type PasswordCheck = {
  issues: PasswordIssue[];
  score: number;
  isValid: boolean;
};

export type PasswordReason =
  | "PASSWORD_TOO_SHORT"
  | "PASSWORD_TOO_LONG"
  | "PASSWORD_TOO_COMMON"
  | "PASSWORD_CONTAINS_EMAIL"
  | "PASSWORD_CONTAINS_NAME"
  | "PASSWORD_NEEDS_VARIETY";

const weakFragments = ["password", "qwerty", "12345678", "letmein", "admin", "playsay", "play-and-say"];

export function checkPassword(password: string, email: string, displayName?: string): PasswordCheck {
  const issues: PasswordIssue[] = [];
  const lowered = password.toLowerCase();

  if (password.length < 8) {
    issues.push("tooShort");
  }
  if (password.length > 128) {
    issues.push("tooLong");
  }
  if (weakFragments.some((fragment) => lowered.includes(fragment))) {
    issues.push("tooCommon");
  }
  const emailFragment = normalizedFragment(email.split("@")[0] ?? "");
  if (emailFragment && lowered.includes(emailFragment)) {
    issues.push("containsEmail");
  }
  if (displayNameFragments(displayName).some((fragment) => lowered.includes(fragment))) {
    issues.push("containsName");
  }
  if (characterClassCount(password) < 3) {
    issues.push("needsVariety");
  }

  const score = Math.max(0, Math.min(4, 4 - issues.length));
  return {
    issues,
    score,
    isValid: issues.length === 0,
  };
}

export function passwordIssueReason(issue: PasswordIssue): PasswordReason {
  return passwordReasonByIssue[issue];
}

function characterClassCount(password: string): number {
  return [
    /\p{Ll}/u.test(password),
    /\p{Lu}/u.test(password),
    /\p{N}/u.test(password),
    /[^\p{L}\p{N}]/u.test(password),
  ].filter(Boolean).length;
}

function normalizedFragment(value: string): string | null {
  const fragment = value.replace(/[^\p{L}\p{N}]/gu, "").toLowerCase();
  return fragment.length >= 3 ? fragment : null;
}

function displayNameFragments(displayName?: string): string[] {
  return (displayName ?? "")
    .split(/[^\p{L}\p{N}]+/u)
    .map((part) => normalizedFragment(part))
    .filter((part): part is string => Boolean(part));
}

const passwordReasonByIssue: Record<PasswordIssue, PasswordReason> = {
  tooShort: "PASSWORD_TOO_SHORT",
  tooLong: "PASSWORD_TOO_LONG",
  tooCommon: "PASSWORD_TOO_COMMON",
  containsEmail: "PASSWORD_CONTAINS_EMAIL",
  containsName: "PASSWORD_CONTAINS_NAME",
  needsVariety: "PASSWORD_NEEDS_VARIETY",
};
